ipc/src/uds.rs

```diff
@@ -0,0 +1,287 @@
+
+pub struct Service<M: Metadata = (), S: Middleware<M> = NoopMiddleware> {
+	handler: Arc<MetaIoHandler<M, S>>,
+	meta: M,
+}
+
+impl<M: Metadata, S: Middleware<M>> Service<M, S> {
+	pub fn new(handler: Arc<MetaIoHandler<M, S>>, meta: M) -> Self {
+		Service { handler: handler, meta: meta }
+	}
+}
+
+impl<M: Metadata, S: Middleware<M>> tokio_service::Service for Service<M, S> {
+	type Request = String;
+	type Response = Option<String>;
+
+	type Error = ();
+
+	type Future = BoxFuture<Self::Response, Self::Error>;
+
+	fn call(&self, req: Self::Request) -> Self::Future {
+		trace!(target: "tcp", "Accepted request: {}", req);
+		self.handler.handle_request(&req, self.meta.clone())
+	}
+}
+
+pub struct ServerBuilder<M: Metadata = (), S: Middleware<M> = NoopMiddleware> {
+	handler: Arc<MetaIoHandler<M, S>>,
+	meta_extractor: Arc<MetaExtractor<M>>,
+	remote: reactor::UnitializedRemote,
+}
+
+impl<M: Metadata, S: Middleware<M> + Send + Sync + 'static> ServerBuilder<M, S> {
+	pub fn new<T>(io_handler: T) -> ServerBuilder<M, S> where
+		T: Into<MetaIoHandler<M, S>>,
+	{
+		Self::with_remote(
+			io_handler,
+			reactor::UnitializedRemote::Unspawned,
+		)
+	}
+
+	pub fn with_remote<T>(
+		io_handler: T,
+		remote: reactor::UnitializedRemote,
+	) -> ServerBuilder<M, S>
+		where T: Into<MetaIoHandler<M, S>>,
+	{
+		ServerBuilder {
+			handler: Arc::new(io_handler.into()),
+			meta_extractor: Arc::new(NoopExtractor),
+			remote: remote,
+		}
+	}	
+
+	/// Run server (in separate thread)
+	pub fn start(self, path: &str) -> std::io::Result<Server> {
+		let remote = self.remote.initialize()?;
+		let rpc_handler = self.handler.clone();		
+		let endpoint_addr = path.to_owned();
+		let meta_extractor = self.meta_extractor.clone();
+		let (stop_signal, stop_receiver) = oneshot::channel();
+		let (start_signal, start_receiver) = oneshot::channel();
+
+		remote.remote().spawn(move |handle| {
+			use self::tokio_uds::UnixListener;
+			use stream_codec::StreamCodec;
+
+			let listener = match UnixListener::bind(&endpoint_addr, handle) {
+				Ok(l) => l,
+				Err(e) => {
+					start_signal.complete(Err(e));
+					return future::ok(()).boxed();
+				}
+			};
+
+			start_signal.complete(Ok(()));
+			let remote = handle.remote().clone();
+			let connections = listener.incoming();
+
+			let server = connections.for_each(move |(unix_stream, client_addr)| {
+				trace!("Accepted incoming UDS connection: {:?}", client_addr);
+
+				let meta = meta_extractor.extract(&RequestContext { endpoint_addr: &client_addr });
+				let service = Service::new(rpc_handler.clone(), meta);
+				let (writer, reader) = unix_stream.framed(StreamCodec).split();
+				let responses = reader.and_then(
+					move |req| service.call(req).then(|response| match response {
+						Err(e) => {
+							warn!(target: "uds", "Error while processing request: {:?}", e);
+							future::ok(String::new())
+						},
+						Ok(None) => {
+							trace!(target: "uds", "JSON RPC request produced no response");
```

Comment:
It's a valid behaviour when notification is received (no `id`) - server should not respond anything in such case.